implementation/transformer/transformer.py

```diff
@@ -0,0 +1,174 @@
+import torch
+
+from torch import nn
+
+
+# The official PyTorch Transformer implementation is missing positional
+# encoding. See https://github.com/pytorch/pytorch/issues/24826. The following
+# implementation of PositionalEncoding seems to be extremely popular.
+# Source: https://pytorch.org/tutorials/beginner/transformer_tutorial
+class PositionalEncoding(nn.Module):
+    def __init__(self, d_model, dropout=0.1, max_len=100):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_ter
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + self.pe[:x.size(0), :]
+        return self.dropout(x)
+
+
+class SimpleTransformerEncoder(nn.Module):
+    """
+    Standard implementation of a Transformer Encoder.
+    Follows:
+        - Original paper diagram: https://arxiv.org/pdf/1706.03762.pdf, Figure 1
+        - Tutorials:
+            - https://pytorch.org/tutorials/beginner/transformer_tutorial.html
+            - https://andrewpeng.dev/transformer-pytorch/
+            - https://nlp.seas.harvard.edu/2018/04/03/attention.html
+    """
+    def __init__(self, vocabulary_size, dropout):
+        super().__init__()
+
+        # TODO: Do we want to use the dropout functionality of
+        # PositionalEncoder, TransformerEncoderLayer, or both?
```

Comment:
I'd leave it for now. In the original paper, they use both but i do not think it has a high priority here.